smallrye-reactive-messaging-mqtt/src/main/java/io/smallrye/reactive/messaging/mqtt/MqttSink.java

```diff
@@ -29,46 +33,118 @@
 
 public class MqttSink {
 
+    private final String channel;
     private final String topic;
     private final int qos;
+    private final boolean healthEnabled;
 
     private final SubscriberBuilder<? extends Message<?>, Void> sink;
+
+    private final AtomicBoolean started = new AtomicBoolean();
     private final AtomicBoolean ready = new AtomicBoolean();
+    private final AtomicReference<Clients.ClientHolder> reference = new AtomicReference<>();
 
     public MqttSink(Vertx vertx, MqttConnectorOutgoingConfiguration config,
             Instance<MqttClientSessionOptions> instances) {
+
         MqttClientSessionOptions options = MqttHelpers.createClientOptions(config, instances);
-        topic = config.getTopic().orElseGet(config::getChannel);
+
+        channel = config.getChannel();
+        topic = config.getTopic().orElse(channel);
         qos = config.getQos();
+        healthEnabled = config.getHealthEnabled();
 
-        AtomicReference<Clients.ClientHolder> reference = new AtomicReference<>();
         sink = ReactiveStreams.<Message<?>> builder()
-                .flatMapCompletionStage(msg -> {
-                    Clients.ClientHolder client = reference.get();
-                    if (client == null) {
-                        client = Clients.getHolder(vertx, options);
-                        reference.set(client);
-                    }
-
-                    return client.start()
-                            .onSuccess(ignore -> ready.set(true))
-                            .map(ignore -> msg)
-                            .toCompletionStage();
-                })
-                .flatMapCompletionStage(msg -> send(reference, msg))
-                .onComplete(() -> {
-                    Clients.ClientHolder c = reference.getAndSet(null);
-                    if (c != null) {
-                        c.close()
-                                .onComplete(ignore -> ready.set(false));
-                    }
-                })
+                .via(new ConnectOnSubscribeProcessor(vertx, options))
+                .flatMapCompletionStage(msg -> send(msg))
                 .onError(log::errorWhileSendingMessageToBroker)
                 .ignore();
+
     }
 
-    private CompletionStage<?> send(AtomicReference<Clients.ClientHolder> reference, Message<?> msg) {
-        MqttClientSession client = reference.get().getClient();
+    /*
+     * This processor let che client mqtt to connect on su
+     */
+    private class ConnectOnSubscribeProcessor implements Processor<Message<?>, Message<?>> {
```

Comment:
I use the lock because could be a race condition between the two sides of processor. Could be possible to have a requestedItem > 0 (line 95) and, on the other side, skip the subscription.request() (line 118), and stay idle forever. If there is only one thread this could not happen, but since I do not sure of that, I've used a lock object.

The difference between ready and started is that started can pass only from false to true when the first connection is done (and stay true), ready can be false if the mqtt connection is down.

I'll remove the sysout.

Reply:
It is tricky to write a processor from scratch. Instead of the lock, the usual way is to check on both `subscribe` and `onSubscribe` and call the `downstream.onSubscribe`  only when you have both downstream and upstream. But what happens if the `client.start` fails ? The implementation also needs to handle cancellation and downstream error propagation. 

I think it is correct to have the startup flag on the subscription time instead of the first outgoing message dispatch.
In the Mutiny Flow API branch, there is an easier way to write that operator using Mutiny `onSubscription`. I'd like to bring that to this branch for easier implementation. Plus it'll eliminate my concerns about conflict. I've pushed the change here if you'd like to integrate to this PR: https://github.com/smallrye/smallrye-reactive-messaging/commit/7636517ecd649c2604ade9998f3bebe556e06200

Otherwise, I still think having both started and ready flags is unnecessary. I would only keep the started flag for the startup probe and use the `client.isConnected()` for both readiness and liveness probes.

Btw that's only for the outgoing channels, for incoming ones what you did should work very well.